Content collections api reference (#2217)
withastro/docs: src/pages/en/reference/api-reference.md

| `page.url.prev`    | `string \| undefined` | Get the URL of the previous page (will be `undefined` if on page 1).                                                              |
| `page.url.next`    | `string \| undefined` | Get the URL of the next page (will be `undefined` if no more pages).                                                              |

## Content Collections (Experimental)

<p>
  <Since v="1.7.0" />
</p>

> Content collections offer APIs to configure and query your Markdown or MDX documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).

Reviewer: ```suggestion
Content collections offer APIs to configure and query your Markdown or MDX documents in `src/content/`. For features and usage examples, [see our content collections guide](/en/guides/content-collections/).
```
nit: Now that the example pulled my attention back up here, I think we can go a little less colloquial on the API reference page.

But, I think.... after that we can update branch?? 🎉

#### `slug`

**Example Type:** `'entry-1' | 'entry-2' | ...`

A URL-ready slug. Defaults to the `id` without the file extension, but can be configured using [the `slug()` config property](#slug). Set to the type `string` if a `slug()` override is configured, and enumerates all possible string values otherwise.

#### `data`

**Type:** `CollectionSchema<TCollectionName>`

An object of frontmatter properties inferred from your collection schema ([see `defineCollection()` reference](#definecollection)). Defaults to `any` if no schema is configured.

#### `body`

**Type:** `string`

A string containing the raw, uncompiled body of the Markdown or MDX document.

#### `render()`

**Type:** `() => Promise<RenderedEntry>`

A function to compile a given Markdown or MDX document for rendering. This returns the following properties:

- `<Content />` - A component used to render the document's contents in an Astro file.
- `headings` - A generated list of headings, [mirroring Astro's `getHeadings()` utility](/en/guides/markdown-content/#exported-properties) on Markdown and MDX imports.

> - `injectedFrontatter` - An object of frontmatter [injected via remark or rehype plugins](/en/guides/markdown-content/#example-injecting-frontmatter). Set to type `any`.

Reviewer 2: ```suggestion
- `injectedFrontmatter ` - An object of frontmatter [injected via remark or rehype plugins](/en/guides/markdown-content/#example-injecting-frontmatter). Set to type `any`.
```

| `page.url.prev`    | `string \| undefined` | Get the URL of the previous page (will be `undefined` if on page 1).                                                              |
| `page.url.next`    | `string \| undefined` | Get the URL of the next page (will be `undefined` if no more pages).                                                              |

## Content Collections (Experimental)

<p>
  <Since v="1.7.0" />
</p>

> Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections).

Reviewer: ```suggestion
Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).
```

> Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).

Author: No prob! Adding "or MDX"

> ### Collection Entry Type

Reviewer: question: Does it maybe make sense to \show these closer to the order in which someone might encounter/use them?

It almost feels like burying the lede a bit to save "defining a collection" to the very end? I might be tempted to go `defineCollection()` first. Then I'd probably think of `getCollection()` and `getEntry()` first, before the the Collection Entry Type? That way, we also don't have the reference of "things below yet to come" in this first line. That is often a sign that maybe things should go in a different order.

I know for this reference page it's not as important as our Guide page in docs. But, I still kind of like organizing these in the order someone might use, and therefore understand, them. Someone *might* be reading these in order, and that seems like a logical flow to me.

Author: Agreed! Reordering to:
1. `defineCollection`
2. `getCollection` and `getEntry`
3. Collection type

```astro
---
import { getEntry } from 'astro:content';
const entry = await getEntry('blog', 'entry-1.md');

const { Content, headings, injectedFrontmatter } = await entry.render();
---
```

See [the rendering entry contents guide](/en/guides/content-collections/#rendering-entry-content) for complete usage examples.

### `getCollection()`

**Type:** `(collection: string, filter?: (entry: CollectionEntry<collection>) => boolean) => CollectionEntry<collection>[]`

`getCollection()` is a function that retrieves a list of content collection entries by collection name.

It returns all items in the collection by default, and accepts an optional `filter` function to narrow by entry properties. This allows you to query for only some items in a collection based on `id`, `slug`, or frontmatter values via the `data` object.

```astro
---
import { getCollection } from 'astro:content';

// Get all `src/content/blog/` entries
const allBlogPosts = await getCollection('blog');

// Only return posts with `draft: true` in the frontmatter
const draftBlogPosts = await getCollection('blog', ({ data }) => {
  return data.draft === true;
});
---
```

> [See the `getCollection()` guide section](/en/guides/content-collections#getcollection) for complete usage examples.

Reviewer: ```suggestion
[See the `getCollection()` guide section](/en/guides/content-collections/#getcollection) for complete usage examples.
```

### `getEntry()`

**Type:** `(collection: string, id: string) => CollectionEntry<collection>`

`getEntry()` is a function that retrieves a single collection entry by collection name and [entry `id`](#id).

```astro
---
import { getEntry } from 'astro:content';

const enterprise = await getEntry('blog', 'enterprise.md');
---
```

### `defineCollection()`

A utility to configure a collection in a `src/content/config.*` file. This is often used when configuring a collection schema.

This function accepts the following properties:

#### `schema`

**Type:** `TSchema extends ZodAnyObject | undefined`

`schema` is an optional object to configure the type and shape of document frontmatter for a collection. Each object value must use [a Zod validator](https://github.com/colinhacks/zod).

#### `slug()`

**Type:** `(entry: { defaultSlug: string } & Omit<CollectionEntry, 'slug'>) => string | undefined`

`slug()` is an optional function to override the default entry slug generated by Astro. This function receives the `defaultSlug`, entry `id`, unparsed document `body`, and `data` parsed by your configured schema, if any.

> #### Usage
> 
> Say you have a collection named `src/content/blog/`, where each Markdown entry has a title and optional `permalink`. You can use `defineCollection` to enforce this frontmatter with the `schema` property. Then, you can map the `permalink` to the entry's `slug` if it exists:

Reviewer: ```suggestion
```
Like all the other times a code sample is used on this page, I would tend not to give this any preamble or explain what it's doing. The guide page shows how to define a collection, enforce frontmatter properties and customize the slug. The code samples here can just be "what it looks like" without much commentary.

I do like the code comment for the export! I think that's sufficient for what needs to go here!

> ```ts

Reviewer: ```suggestion
```ts
// src/content/config.ts
import { z, defineCollection } from 'astro:content';
```
We don't show importing `defineCollection()` here, though we do show exporting it. Is this helpful?
